syntax = "proto3";

package json_logger;

import "google/protobuf/timestamp.proto";

service JsonLogger {
  // Initializes the given JSON file for logging.
  // Status Codes for errors:
  // - INVALID_ARGUMENT: Invalid JSON File or Invalid Session Initialization Behavior.
  // - PERMISSION_DENIED: Permission denied for the File.
  // - INTERNAL: File path is invalid or inaccessible or any other unexpected behavior.
  // - ALREADY_EXISTS: File has already been initialized and cannot be initialized again for SESSION_INITIALIZATION_BEHAVIOR_INITIALIZE_NEW.
  // - NOT_FOUND: Session does not exist for SESSION_INITIALIZATION_BEHAVIOR_ATTACH_TO_EXISTING.
  rpc InitializeFile(InitializeFileRequest) returns (InitializeFileResponse);

  // Logs measurement data to the file of the session.
  // Status Codes for errors:
  // - PERMISSION_DENIED: Permission denied for the File.
  // - NOT_FOUND: Session does not exist.
  // - INTERNAL: File path is invalid or inaccessible or any other unexpected behavior.
  rpc LogMeasurementData(LogMeasurementDataRequest) returns (LogMeasurementDataResponse);

  // Closes the file handle of the session.
  // Status Codes for errors:
  // - NOT_FOUND: Session does not exist.
  // - INTERNAL: Any unexpected behavior.
  rpc CloseFile(CloseFileRequest) returns (CloseFileResponse);
}

message InitializeFileRequest {
  // The complete file path.
  string file_path = 1;
  // Defines how the session should be initialized.
  SessionInitializationBehavior initialization_behavior = 2;
}

message InitializeFileResponse {
  string session_name = 1;
  bool new_session = 2;
}

message LogMeasurementDataRequest{
  string session_name = 1;
  string measurement_name = 2;
  google.protobuf.Timestamp timestamp = 3;
  map<string, string> measurement_configurations = 4;
  map<string, string> measurement_outputs=5;
}

// Empty response message; success is indicated by the gRPC status code.
message LogMeasurementDataResponse {
}

message CloseFileRequest {
  string session_name = 1;
}

message CloseFileResponse {
}

enum SessionInitializationBehavior {
  // Automatically decide whether to initialize new or attach.
  SESSION_INITIALIZATION_BEHAVIOR_UNSPECIFIED = 0;
  SESSION_INITIALIZATION_BEHAVIOR_INITIALIZE_NEW = 1;
  SESSION_INITIALIZATION_BEHAVIOR_ATTACH_TO_EXISTING = 2;
}
